packages/back-end/src/util/legacyAudit/maps.ts

```diff
@@ -0,0 +1,169 @@
+import { UnionToIntersection } from "../types";
+import { DefinedEvent } from "../../events/notification-events";
+import { NotificationEventTemplate } from "../../events/base-types";
+import {
+  AuditNotificationEventMap,
+  AuditEventNameTemplate,
+  AuditEventResource,
+  auditNotificationEvent,
+  auditNotificationEvents,
+  auditEvents,
+} from "./base";
+
+type AuditEventMappingTemplate<
+  R,
+  E = AuditEventNameTemplate<R>
+> = R extends AuditEventResource
+  ? E extends AuditEventNameTemplate<R>
+    ? // Here we exclude the audit events that are already defined as notification event.
+      `${R}.${AuditNotificationEventMap<R, E>}` extends DefinedEvent
+      ? never
+      : {
+          [k in `${R}.${E}`]: {
+            object: R;
+            event: `${R}.${AuditNotificationEventMap<R, E>}`;
+            auditData: {
+              id: string;
+              reason?: string;
+              parent?: {
+                object: R;
+                id: string;
+              };
+              details?: string;
+            };
+          };
+        }
+    : never
+  : never;
+
+// This is used to mapaudit interface to event interface.
+// It has type: { "experiment.update": { entity: "experiment", event: "experiment.updated", ... } & ...
+// and excludes the events already defined.
+type AuditEventMappings = UnionToIntersection<
+  AuditEventMappingTemplate<AuditEventResource>
+>;
+
+export const auditEventMappings = ({
```

Comment:
The type of this function is forced. It should be proof-read to make sure it actually implements what it says it does.